internal/translation/ipaccesslist/conversion.go

```diff
@@ -0,0 +1,182 @@
+package ipaccesslist
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"go.mongodb.org/atlas-sdk/v20231115008/admin"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/timeutil"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/project"
+)
+
+type IPAccessEntry struct {
+	CIDR             string
+	AWSSecurityGroup string
+	DeleteAfterDate  *time.Time
+	Comment          string
+}
+
+func (i *IPAccessEntry) ID() string {
+	if i.CIDR != "" {
+		return i.CIDR
+	}
+
+	return i.AWSSecurityGroup
+}
+
+func (i *IPAccessEntry) IsExpired(at time.Time) bool {
+	if i.DeleteAfterDate == nil {
+		return false
+	}
+
+	return i.DeleteAfterDate.Before(at)
+}
+
+type IPAccessEntries map[string]*IPAccessEntry
+
+func (i IPAccessEntries) GetByStatus(expired bool) IPAccessEntries {
+	entries := make(IPAccessEntries, len(i))
+
+	for ix, entry := range i {
+		if entry.IsExpired(time.Now()) == expired {
+			entries[ix] = entry
+		}
+	}
+
+	return entries
+}
+
+func FromInternal(ipAccessEntries IPAccessEntries) []project.IPAccessList {
```

Comment:
I was not understanding why we needed this, but then I saw we are changing the shape of the IP access list from an array to a set. That is fine, I guess it simplifies something later like checking for no duplicates.

Still most other "internal" types augment information that was not there in the spec, such as the project id or a password value.

To avoid the need for this `FromInternal` you could use the composition trick again:

```go
type IPAccessEntries struct {
  List []project.IPAccessList
  Set map[string]*IPAccessEntry
}
```

When passing the type you just need to use the `.List` instead of calling `FromInternal`. It is the original list, so no bugs possible.

Reply:
I saw later that this is used after a filtering call, so the trick will not work.